kani-compiler/src/kani_middle/transform/kani_intrinsics.rs

```diff
@@ -120,11 +127,141 @@ impl IntrinsicGeneratorPass {
-                let result = new_body.new_assignment(rvalue, &mut terminator);
+                let result =
+                    new_body.new_assignment(rvalue, &mut terminator, InsertPosition::Before);
                 let reason = format!(
                     "Kani currently doesn't support checking validity of `{target_ty}`. {msg}"
                 );
-                new_body.add_check(tcx, &self.check_type, &mut terminator, result, &reason);
+                new_body.add_check(
+                    tcx,
+                    &self.check_type,
+                    &mut terminator,
+                    InsertPosition::Before,
+                    result,
+                    &reason,
+                );
+            }
+        }
+        new_body.into()
+    }
+
+    /// Generate the body for `is_initialized`, which looks like the following
+    ///
+    /// ```
+    /// pub fn is_initialized<T>(ptr: *const T, len: usize) -> bool {
+    ///     let layout = ... // Byte mask representing the layout of T.
+    ///     __kani_mem_init_sm_get(ptr, layout, len)
+    /// }
+    /// ```
+    fn is_initialized_body(&self, tcx: TyCtxt, body: Body) -> Body {
+        let mut new_body = MutableBody::from(body);
+        new_body.clear_body();
+
+        // Initialize return variable with True.
+        let ret_var = RETURN_LOCAL;
+        let mut terminator = SourceInstruction::Terminator { bb: 0 };
+        let span = new_body.locals()[ret_var].span;
+        let assign = StatementKind::Assign(
+            Place::from(ret_var),
+            Rvalue::Use(Operand::Constant(ConstOperand {
+                span,
+                user_ty: None,
+                const_: MirConst::from_bool(true),
+            })),
+        );
+        let stmt = Statement { kind: assign, span };
+        new_body.insert_stmt(stmt, &mut terminator, InsertPosition::Before);
+
+        // The first argument type.
+        let arg_ty = new_body.locals()[1].ty;
+        let TyKind::RigidTy(RigidTy::RawPtr(target_ty, _)) = arg_ty.kind() else { unreachable!() };
+        let pointee_info = PointeeInfo::from_ty(target_ty);
+        match pointee_info {
+            Ok(pointee_info) => {
+                match pointee_info.layout() {
+                    PointeeLayout::Static { layout } => {
+                        let shadow_memory_get_instance = Instance::resolve(
+                            get_kani_sm_function(tcx, "KaniMemInitSMGet"),
+                            &GenericArgs(vec![
+                                GenericArgKind::Const(
+                                    TyConst::try_from_target_usize(
+                                        layout.to_byte_mask().len() as u64
+                                    )
+                                    .unwrap(),
+                                ),
+                                GenericArgKind::Type(*pointee_info.ty()),
+                            ]),
+                        )
+                        .unwrap();
```

Comment:
I think every time you call `get_kani_sm_function` you are trying to resolve an instance and this chunk of code is replicated. Can you create a function that takes the `FnDef`, a `usize` and a `Ty` returns an instance instead?